cli/compiler.rs

```diff
@@ -74,229 +96,552 @@ fn req(
+    let compiler_rid = resource.rid;
+    let first_msg_fut =
+      resources::post_message_to_worker(compiler_rid, req_msg)
+        .then(move |_| worker)
+        .then(move |result| {
+          if let Err(err) = result {
+            // TODO(ry) Need to forward the error instead of exiting.
+            eprintln!("{}", err.to_string());
+            std::process::exit(1);
+          }
+          debug!("Sent message to worker");
+          let stream_future =
+            resources::get_message_stream_from_worker(compiler_rid)
+              .into_future();
+          stream_future.map(|(f, _rest)| f).map_err(|(f, _rest)| f)
+        });
+
+    first_msg_fut.map_err(|_| panic!("not handled")).and_then(
+      move |maybe_msg: Option<Buf>| {
+        debug!("Received message from worker");
+
+        if let Some(msg) = maybe_msg {
+          let json_str = std::str::from_utf8(&msg).unwrap();
+          debug!("Message: {}", json_str);
+          if let Some(diagnostics) = Diagnostic::from_emit_result(json_str) {
+            return Err(ErrBox::from(diagnostics));
+          }
         }
-      }
 
-      Ok(())
-    },
-  )
-}
+        Ok(())
+      },
+    )
+  }
+
+  fn mark_compiled(&self, module_id: &str) {
+    let mut c = self.compiled.lock().unwrap();
+    c.insert(module_id.to_string());
+  }
 
-pub fn compile_async(
-  state: ThreadSafeState,
-  module_meta_data: &ModuleMetaData,
-) -> impl Future<Item = ModuleMetaData, Error = ErrBox> {
-  let module_name = module_meta_data.module_name.clone();
-
-  debug!(
-    "Running rust part of compile_sync. module_name: {}",
-    &module_name
-  );
-
-  let root_names = vec![module_name.clone()];
-  let compiler_config = get_compiler_config(&state, "typescript");
-  let req_msg = req(root_names, compiler_config, None);
-
-  // Count how many times we start the compiler worker.
-  state.metrics.compiler_starts.fetch_add(1, Ordering::SeqCst);
-
-  let mut worker = Worker::new(
-    "TS".to_string(),
-    startup_data::compiler_isolate_init(),
-    // TODO(ry) Maybe we should use a separate state for the compiler.
-    // as was done previously.
-    state.clone(),
-  );
-  worker.execute("denoMain()").unwrap();
-  worker.execute("workerMain()").unwrap();
-  worker.execute("compilerMain()").unwrap();
-
-  let compiling_job = state.progress.add("Compile", &module_name);
-
-  let resource = worker.state.resource.clone();
-  let compiler_rid = resource.rid;
-  let first_msg_fut = resources::post_message_to_worker(compiler_rid, req_msg)
-    .then(move |_| worker)
-    .then(move |result| {
-      if let Err(err) = result {
-        // TODO(ry) Need to forward the error instead of exiting.
-        eprintln!("{}", err.to_string());
-        std::process::exit(1);
+  fn has_compiled(&self, module_id: &str) -> bool {
+    let c = self.compiled.lock().unwrap();
+    c.contains(module_id)
+  }
+
+  pub fn compile_async(
+    self: &Self,
+    state: ThreadSafeState,
+    source_file: &SourceFile,
```

Comment:
It's odd that this function takes a source_file argument rather than a URL.

Reply:
Or should it be `specifier: &ModuleSpecifier` ?

Reply:
Yeah I still haven't settled on API... A few lines below there's TODO about fetching `SourceFile` inside `compiler_async`. I'll leave it as is, and figure out compiler pipeline in followup PR